sys/linux/init_images.go

```diff
@@ -0,0 +1,243 @@
+		if idx > 0 {
+			// Adjust the end of the previous segment.
+			prevSegment := parsed.segments[idx-1]
+			if prevSegment.offset.Val+prevSegment.size.Val > segment.offset.Val {
+				resizeSegment(prevSegment, segment.offset.Val-prevSegment.offset.Val)
+			}
+		}
+		if segment.offset.Val+segment.size.Val > imageMaxSize {
+			resizeSegment(segment, imageMaxSize-segment.offset.Val)
+		}
+		if segment.offset.Val+segment.size.Val > newSize {
+			newSize = segment.offset.Val + segment.size.Val
+		}
+	}
+	if newSize > imageMaxSize {
+		newSize = imageMaxSize
+	}
+	parsed.size.Val = newSize
+
+	// Drop 0-size segments.
+	parsed.filterSegments(func(i int, segment *mountImageSegment) bool {
+		return segment.size.Val > 0
+	})
+}
+
+type sortSegments struct {
+	parsed *mountImageArgs
+}
+
+func (s sortSegments) Len() int { return len(s.parsed.segments) }
+func (s sortSegments) Swap(i, j int) {
+	inner := s.parsed.segmentsGroup.Inner
+	inner[i], inner[j] = inner[j], inner[i]
+	s.parsed.segments[i], s.parsed.segments[j] = s.parsed.segments[j], s.parsed.segments[i]
+}
+func (s sortSegments) Less(i, j int) bool {
+	segments := s.parsed.segments
+	if segments[i].offset.Val != segments[j].offset.Val {
+		return segments[i].offset.Val < segments[j].offset.Val
+	}
+	return segments[i].size.Val < segments[j].size.Val
+}
+
+func (arch *arch) fixUpSyzMountImage(c *prog.Call) {
+	// Previously we did such a sanitization right in the common_linux.h, but this was problematic
+	// for two reasons:
+	// 1) It further complicates the already complicated executor code.
+	// 2) We'd need to duplicate the logic in Go for raw image extraction.
+	// So now we do all the initialization in Go and let the C code only interpret the commands.
+	ret, err := parseSyzMountImage(c)
+	if err != nil {
+		deactivateSyzMountImage(c)
+		return
+	}
+	const maxImageSegments = 4096
+	ret.filterSegments(func(i int, _ *mountImageSegment) bool {
+		return i < maxImageSegments
+	})
+	fixUpImageSegments(ret)
+}
+
+type mountImageArgs struct {
+	size          *prog.ConstArg
+	segmentsCount *prog.ConstArg
+	segmentsGroup *prog.GroupArg
+	segments      []*mountImageSegment
+}
+
+func (m *mountImageArgs) filterSegments(filter func(int, *mountImageSegment) bool) {
+	newArgs := []prog.Arg{}
+	newSegments := []*mountImageSegment{}
+	for i, segment := range m.segments {
+		if filter(i, segment) {
+			newSegments = append(newSegments, segment)
+			newArgs = append(newArgs, m.segmentsGroup.Inner[i])
+		}
+	}
+	m.segments = newSegments
+	m.segmentsGroup.Inner = newArgs
+	m.segmentsCount.Val = uint64(len(newArgs))
+}
+
+type mountImageSegment struct {
+	data       *prog.DataArg
+	offset     *prog.ConstArg
+	size       *prog.ConstArg
+	parseError error
+}
+
+func parseImageSegment(segmentArg prog.Arg) *mountImageSegment {
+	ret := &mountImageSegment{}
+	segmentFields, ok := segmentArg.(*prog.GroupArg)
+	if segmentFields == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("it is not a group")}
```

Comment:
I think it's better to panic in these cases.
Otherwise image extraction/sanitization will silently stop working and we won't notice for a long time.

Reply:
Is it not possible that during mutation syzkaller could replace the group arg with something else? (E.g. squash it to some blob)
And same question to other comments below about nils, etc. Is it guaranteed that the types of arguments/subarguments/etc will always correspond to those of the original call descriptions?

Reply:
> E.g. squash it to some blob

Not sure. There are some heuristics to decide what types to squash.
But if it's to deal with ANY types, this is not an error. You can check for ANY type and return nil image.

> Is it guaranteed that the types of arguments/subarguments/etc will always correspond to those of the original call descriptions?

Yes, they should.

A test that generates syz_mount_image, then mutates and then extracts image would catch all of these conditions.

Reply:
I've changed almost all those errors to panic and written a test that mutates `syz_mount_image`. So far none of the panics were triggered.

But I'm still worried about the ANY transformations -- by looking at the code I cannot prove that the segments will never be squashed. So I left the segment argument handling as it was and also left the deactivation code in place -- to handle such cases.

Reply:
My test did not fail, but one of the existing ones did:

```
--- FAIL: TestDefaultCallArgs (8.60s)
    --- FAIL: TestDefaultCallArgs/linux/s390x (0.14s)
panic: syz_mount_image's segments arg is not ptr [recovered]
	panic: syz_mount_image's segments arg is not ptr
target: linux/s390x, rev: , mode=1, prog:
"syz_mount_image$fuse()" [recovered]
	panic: syz_mount_image's segments arg is not ptr
target: linux/s390x, rev: , mode=1, prog:
"syz_mount_image$fuse()"
```

Reply:
Fixed that by reverting that particular panic to a normal error again.